api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionDefineNetworkChannelsEvent.java

```diff
@@ -0,0 +1,217 @@
+import org.geysermc.geyser.api.network.message.Message;
+import org.geysermc.geyser.api.network.message.MessageBuffer;
+import org.geysermc.geyser.api.network.message.MessageCodec;
+import org.geysermc.geyser.api.network.message.MessageFactory;
+import org.geysermc.geyser.api.network.message.MessageHandler;
+import org.geysermc.geyser.api.network.message.MessagePriority;
+
+import java.util.function.Consumer;
+
+/**
+ * Called whenever Geyser is registering network channels.
+ * @since 2.9.1
+ */
+public abstract class SessionDefineNetworkChannelsEvent extends ConnectionEvent {
+
+    public SessionDefineNetworkChannelsEvent(@NonNull GeyserConnection connection) {
+        super(connection);
+    }
+
+    /**
+     * Defines the registration of a new network channel with a message factory.
+     *
+     * @param channel the channel to register
+     * @param messageFactory the factory to create messages from the buffer
+     * @param <M> the message type created by the factory
+     * @return a registration builder to configure handlers
+     */
+    public abstract <M extends Message<MessageBuffer>> Builder.@NonNull Initial<M> define(@NonNull NetworkChannel channel, @NonNull MessageFactory<MessageBuffer, M> messageFactory);
+
+    /**
+     * Defines the registration of a new network channel with a codec and message factory.
+     *
+     * @param channel the channel to register
+     * @param codec the codec to use to encode/decode the buffer
+     * @param messageFactory the factory to create messages from the buffer
+     * @param <T> the buffer type
+     * @param <M> the message type created by the factory
+     * @return a registration builder to configure handlers
+     */
+    public abstract <T extends MessageBuffer, M extends Message<T>> Builder.@NonNull Initial<M> define(@NonNull NetworkChannel channel, @NonNull MessageCodec<T> codec, @NonNull MessageFactory<T, M> messageFactory);
+
+    /**
+     * Registration builder for attaching handlers to a channel.
+     *
+     * @param <M> the message type
+     */
+    public interface Builder<M extends Message<? extends MessageBuffer>> {
+
+        /**
+         * Configures the pipeline for this handler.
+         *
+         * @param pipeline the pipeline consumer
+         * @return the builder instance
+         */
+        @NonNull
+        Builder<M> pipeline(@NonNull Consumer<Pipeline> pipeline);
+
+        /**
+         * Finalizes the registration.
+         *
+         * @return the completed registration
+         */
+        @NonNull
+        Registration<M> register();
+
+        interface Initial<M extends Message<? extends MessageBuffer>> extends Sided<M>, Bidirectional<M> {
+
+            /**
+             * Sets the protocol state for this message.
+             *
+             * @param state the protocol state
+             * @return the initial builder instance
+             */
+            @NonNull
+            Initial<M> protocolState(@NonNull ProtocolState state);
```

Comment:
This is a bit unclear - what about Bedrock packets; they don't have protocol states? Or plugin messages, that can be sent in (almost) any stage?
It might be good to be able to specify multiple valid protocol states, although that'd be somewhat niche

Reply:
Multiple protocol states would not be necessary, as the packet ID will (in nearly every case) change depending on the state.